stm32-modules/include/flex-stacker/flex-stacker/tmf8821.hpp

```diff
@@ -0,0 +1,457 @@
+        auto reg = registers.enable;
+        reg.padding_1 = 0;
+        return set_register(reg, sensor_id).has_value();
+    }
+
+    auto write(TOFSensorID sensor_id, uint16_t reg, uint32_t* data,
+               int size = 1) -> std::optional<RegisterSerializedType> {
+        using RT = std::optional<RegisterSerializedType>;
+        auto dev_address = get_sensor_i2c_address(sensor_id);
+        auto [res, _] =
+            _policy->i2c_write(dev_address << 1, reg, (uint8_t*)data, size);
+        if (res != 0) return RT();
+        return RT(res);
+    }
+
+    auto read(TOFSensorID sensor_id, uint32_t reg, int size = 1)
+        -> std::optional<RegisterSerializedType> {
+        using RT = std::optional<RegisterSerializedType>;
+        auto dev_address = get_sensor_i2c_address(sensor_id);
+        auto [res, data] = _policy->i2c_read(dev_address << 1, reg, size);
+        if (res != 0) return RT();
+        auto value = static_cast<uint32_t>(*data.data());
+        return RT(value);
+    }
+
+    /* Check which app (bootloader or measurement ) is running. */
+    auto get_sensor_mode(TOFSensorID sensor_id) -> TOFSensorMode {
+        auto ret = read_register<tmf8821::AppID>(sensor_id);
+        if (!ret.has_value()) {
+            return TOFSensorMode::UNKNOWN;
+        }
+        auto appid = static_cast<tmf8821::AppID>(ret.value()).appid;
+        return static_cast<TOFSensorMode>(appid);
+    }
+
+    auto reset_custom_address() -> void { _custom_address = false; }
+
+    // Gets the sensor i2c address
+    auto get_sensor_i2c_address(TOFSensorID sensor_id, bool default_addr = true)
+        -> uint16_t {
+        if (default_addr && !_custom_address) return TOF_DEFAULT_ADDRESS;
+        if (sensor_id == TOF_X) return TOF_X_ADDRESS;
+        if (sensor_id == TOF_Z) return TOF_Z_ADDRESS;
+        return TOF_DEFAULT_ADDRESS;
+    }
+
+    auto set_sensor_i2c_address(TOFSensorID sensor_id, uint8_t address)
+        -> bool {
+        // switch to the appid=0x03, cid_rid=0x16 – Configuration Page
+        if (!change_config_page(sensor_id, CMD_LOAD_CONFIG_PAGE_COMMON)) {
+            return false;
+        }
+
+        // Verify page change
+        if (!wait_for_state(sensor_id, _registers.cfg_result.address,
+                            CMD_LOAD_CONFIG_PAGE_COMMON)) {
+            return false;
+        }
+
+        // Set the new address in the I2C_SLAVE_ADDRESS (0x3B) register.
+        _registers.i2c_address.slave_address = address;
+        if (!set_register(_registers.i2c_address, sensor_id).has_value()) {
+            return false;
+        }
+
+        // Set change address I2C_ADDR_CHANGE (0x3E) to 0
+        if (!set_register(_registers.i2c_addr_change, sensor_id).has_value()) {
+            return false;
+        }
+
+        // Write the config page
+        auto len = prepare_cmd_frame(CMD_WRITE_CONFIG_PAGE, 0, 0);
+        if (!write(sensor_id, _registers.cmd_stat.address, (uint32_t*)BUFFER,
```

Comment:
if this is what we're using `BUFFER` for then it definitely should be a class member. There's definitely going to be 2 of these class instances, which means that they could stomp on each other if they're sharing the same buffer. Putting it in the class will avoid that.

Reply:
also the buffer size is defined in bytes and is `131` which definitely does not align with punning it to an array of uint32s, we'd get a guaranteed overflow if we ever wrote to that final word